pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -204,6 +222,48 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileNodeMaintenance) checkQuorumValidity(node *corev1.Node) (bool, error) {
+
+	if !isMasterNode(node) {
+		log.Infof("node %s is not a master", node.Name)
+		return true, nil
+	}
+
+	masterNodes, availableMasterNodes, err := r.countMasterNodes()
+	if err != nil {
+		return false, err
+	}
+
+	quorumValid := availableMasterNodes-1 >= (masterNodes/2 + 1)
+
+	log.Infof("Node %s is a master, but quorum is still maintained. (availableMasters %d totalMasters %d", node.Name, availableMasterNodes, masterNodes)
+
+	return quorumValid, nil
+}
+
+func (r *ReconcileNodeMaintenance) countMasterNodes() (int, int, error) {
+	masterNodes := 0
+	availableMasterNodes := 0
+
+	n, err := r.drainer.Client.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: MasterNodeLabelSelector})
```

Comment:
That is interesting. I would like to learn more about that. The bugzilla report for this PR (link at the top) advised me to look at https://github.com/openshift/etcd-quorum-guard - the e2e test in this repository were all counting the master nodes by listing all nodes with that label selector.

Reply:
When baremetal Machine turns unhealthy, we shut it down (using hardware management interface), remove the node from the cluster, and power it back on.
It makes me think that we need to be more careful with masters remediation. It feels like MCO/NMO might drain a master node, while machine-remediation might take other master down. 
/cc @beekhof

Reply:
How do you remove the node from the cluster? I understand from your previous comment that it is still returned by the go-client List call, How can I know if a node has been 'removed from the cluster' ?

Reply:
See implementation [here](https://github.com/openshift/cluster-api-provider-baremetal/blob/master/pkg/cloud/baremetal/actuators/machine/actuator.go#L952).

I don't expect it to be returned by the client.

Reply:
ok, so once you delete the node object via client-go and it gets deleted from etcd, then it will not be picked up by the List selector, so I don't see a problem to begin with.

Reply:
Perhaps we could add a placeholder object upon deleting a master node, and delete the placeholder object when the deleted master node comes up again, then I could count the placeholder objects and add it to the number of active master nodes. We could also avoid conflicts by storing the node name in the placeholder object, so that nodes will not be counted twice.

Reply:
I think the optimal approach would be getting this from etcd-operator or any other entity in the system that has the knowledge. Not sure if this is possible though

Reply:
They gather this knowledge from the raft consensus algorithm. I am not sure that a simple operator wants to be exposed to the internals of the etcd raft implementation; these things might change suddenly between versions of opernshift/kubernetes and we will be the last to know they changed.

Reply:
now the PR is checking for the MCO PDB. if that doesn't exist then we still use this code as a fallback option.